unyt/array.py

```diff
@@ -1376,30 +1378,37 @@ def from_string(s, unit_registry=None):
         --------
         >>> from unyt import unyt_quantity
         >>> unyt_quantity.from_string("1cm")
-        unyt_quantity(1., 'cm')
+        unyt_quantity(1, 'cm')
         >>> unyt_quantity.from_string("+1e3 Mearth")
         unyt_quantity(1000., 'Mearth')
         >>> unyt_quantity.from_string("-10. kg")
         unyt_quantity(-10., 'kg')
         >>> unyt_quantity.from_string(".66\tum")
         unyt_quantity(0.66, 'μm')
         >>> unyt_quantity.from_string("42")
-        unyt_quantity(42., '(dimensionless)')
+        unyt_quantity(42, '(dimensionless)')
         >>> unyt_quantity.from_string("1.0 g/cm**3")
         unyt_quantity(1., 'g/cm**3')
         """
         v = s.strip()
         if re.fullmatch(_NUMB_REGEXP, v):
-            return float(re.match(_NUMB_REGEXP, v).group()) * Unit()
-        if re.fullmatch(_UNIT_REGEXP, v):
-            return 1 * Unit(re.match(_UNIT_REGEXP, v).group())
-        if not re.match(_QUAN_REGEXP, v):
+            num = re.match(_NUMB_REGEXP, v).group()
+            unit = Unit()
+        elif re.fullmatch(_UNIT_REGEXP, v):
+            num = 1
+            unit = Unit(re.match(_UNIT_REGEXP, v).group())
+        elif not re.match(_QUAN_REGEXP, v):
             raise ValueError(f"Received invalid quantity expression '{s}'.")
-        res = re.search(_NUMB_REGEXP, v)
-        num = res.group()
-        res = re.search(_UNIT_REGEXP, v[res.span()[1] :])
-        unit = res.group()
-        return float(num) * Unit(unit, registry=unit_registry)
+        else:
+            res = re.search(_NUMB_REGEXP, v)
+            num = res.group()
+            res = re.search(_UNIT_REGEXP, v[res.span()[1] :])
+            unit = res.group()
```

Comment:
This is a nit, I'm trying to avoid reusing variable names for different types so we can maybe start doing static type-checking without too much rewrites in the future
```suggestion
            snum = re.search(_NUMB_REGEXP, v).group()
            sunit = re.search(_UNIT_REGEXP, v[res.span()[1] :]).group()
```